trustfall_core/src/interpreter/hints/vertex_info.rs

```diff
@@ -23,6 +34,8 @@ pub trait VertexInfo: super::sealed::__Sealed {
     /// The type coercion (`... on SomeType`) applied by the query at this vertex, if any.
     fn coerced_to_type(&self) -> Option<&Arc<str>>;
 
+    fn required_properties(&self) -> Vec<RequiredProperty>;
```

Comment:
Is this a good signature/name?

Reply:
> Is Vec or impl Iterator better?

I think `impl Iterator` would be better here since I don't think the caller needs a specific data structure besides the ability to iterate over it. But I was unsure about it because I will have to add some indirection (Box<dyn object>) since I think we don't want this to be generic, right? Should I change the signature to return `Box<dyn Iterator<Item=RequiredProperty>>`?

> Can the same property appear more than once?
 My understanding of the IR is that it cannot, but I'm not enforcing it. Do you think is worth having a `HashSet` here?
 
 > Is there a guaranteed order in which elements appear?

If I understood correctly, it should because "Trustfall is a deterministic compiler; any non-determinism is considered a bug
and will be caught by tests.". Though I'm not sure if this is meant to be just the output of the compiler (the IR) or everything inside Trustfall. 

But given that we are iterating over a `BTree` for all properties and we are using a `Vec`, we should always produce the same output for the same query.

> Are these only the properties from "this" vertex (the one whose VertexInfo this is, as evidenced by the .vid() accessor), or are they properties from more than one vertex?

I believe it should be only from "this" vertex. Although for graphql it may make sense to also get the properties from the "nested" vertex I thought it would be weird to do such a thing, since my understanding of trustfall is that those vertexes will also be "visited" and the adapter will have to fetch that data.

Reply:
> What happens if the same property has both an output and a filter? Or also a tag?

Ah, good point! I will have to handle the duplicates!